feat(ingest/spark): upgrade OpenLineage to 1.50.0 with full shadow-jar shading + Spark 4 support

treff7es opened the pull request:
## Summary

Upgrades the DataHub Spark lineage agent's OpenLineage integration to **1.50.0** and
verifies **Apache Spark 4.x** support. This PR now carries the full upgrade in one
place (it absorbs the follow-up that was tracked separately in #16890, which is
closed as superseded).

## What's included

- **OpenLineage → 1.50.0** (from the master baseline), with the vendored/shaded
  OpenLineage classes refreshed onto the upstream 1.50 base per `CLAUDE.md`, and the
  DataHub customization patches regenerated (`v1.50.0`).
- **Per-connection `platform_instance` Glue fix**: OL 1.46 (#4384, `isIcebergUsingGlue`)
  ships natively, so the temporary DataHub `AwsUtils` backport was removed. Verified by
  `SparkGlueConnectionInstanceSmokeTest`.
- **OL 1.48 `DatasetFactory` builder-pattern migration** across the vendored classes;
  reflection-guarded the Spark-version-specific `RddDatasetInfoExtractor` branches
  (`SparkPath` / `inputPartitions()` renames).
- **Full shadow-jar relocations** to prevent classpath conflicts on Databricks/EMR/Glue.
- **Apache Spark 4.0.0 compatibility smoke test** (`sparkSmoke4Test`, Scala 2.13) running
  against the real shaded agent jar, wired into the `spark-smoke-test` CI workflow.
- Lockfiles regenerated; `CLAUDE.md` upgrade docs updated.

## Verification

`spark-smoke-test` CI is green: unit tests, the real-Spark smoke suite (incl. the Glue
connection-instance test), the Spark 4 compatibility test, and `checkShadowJar` all pass.

treff7es committed:
Map sqlserver to mssql in dialect detection

treff7es committed:
Initial commit for openlineage 1_38 upgrade

treff7es committed:
Add proper shading

treff7es committed:
Fix doc formatting

treff7es committed:
fix(ingest/spark): prevent empty dataJobInputOutput UPSERT from clobbering output lineage edges

When coalesced emission fires on early Spark events (e.g., START), the
dataJobInputOutput aspect was emitted with empty inputDatasetEdges and
outputDatasetEdges arrays via UPSERT. Later PATCH emissions that added
edges could not reliably override this empty state due to async processing
ordering, causing output dataset lineage to be silently lost.

Fixes:
- Skip dataJobInputOutput emission when all edge arrays are empty
- Null-check SparkEnv.get() in PlanUtils.getDirectoryPath to prevent NPE
  during Spark shutdown, falling back to original path resolution
- Handle URISyntaxException gracefully in RemovePathPatternUtils instead
  of wrapping in RuntimeException (crashes the lineage pipeline)
- Fix double-Optional wrapping in RemovePathPatternUtils.getPattern()
  that threw NoSuchElementException when SparkContext was absent
- Guard JDBC option access in SaveIntoDataSourceCommandVisitor to prevent
  NoSuchElementException for query-based JDBC sinks
- Reuse built MCP in patch path instead of calling build() twice
- Downgrade per-dataset log.info to log.debug in RemovePathPatternUtils

treff7es committed:
docs(ingest/spark): add changelog entries for output lineage and stability fixes

treff7es committed:
docs(ingest/spark): add OpenLineage 1.38→1.45 upgrade research document

treff7es committed:
feat(ingest/spark): upgrade OpenLineage from 1.38.0 to 1.45.0

Upgrades the OpenLineage Spark integration dependency from 1.38.0 to 1.45.0.

Changes:
- Updated ext.openLineageVersion to 1.45.0 in root build.gradle
- Copied upstream 1.45.0 source files and reapplied all DataHub patches
- Handled RddPathUtils.java → RddDatasetInfoExtractor.java rename
  (upstream expanded from 225 to 362 lines, DataHub's 3 log-level
  changes reapplied to new locations)
- Preserved upstream's new CLL facet cleanup in RemovePathPatternUtils
  (PR #4228 — path pattern removal now applies to ColumnLineageFacet)
- Incorporated upstream's S3 getFileStatus optimization (PR #3985),
  DataSourceRDD symlink extraction (PR #4283), and Iceberg schema
  support (PR #4331)

All 9 DataHub patches applied cleanly against 1.45.0 upstream.
The only manual work was the RddPathUtils rename and fixing imports
where upstream removed them but DataHub's patches still need them.

treff7es committed:
chore(ingest/spark): generate v1.45.0 patch files and fix generate-patches script

- Generated 11 patch files for v1.45.0 in patches/datahub-customizations/v1.45.0/
- Fixed generate-patches.sh to output to versioned subdirectory (v<version>/)
- Updated file list: RddPathUtils.java → RddDatasetInfoExtractor.java

treff7es committed:
fix(ingest/spark): relocate additional packages to prevent Spark classpath conflicts

Adds shadow JAR relocations for packages that were shipping unrelocated
and could conflict with Spark's runtime classpath:

- org.apache.kafka (3765 classes) — conflicts with Spark structured streaming
- org.apache.log4j (217 classes) — conflicts with Spark's logging
- org.apache.http (723 classes) — httpcomponents 4.x not caught by existing
  org.apache.hc rule (which only covers httpclient5)
- org.apache.commons.text, com.google.re2j, org.eclipse.parsson,
  org.codehaus.plexus, org.apache.maven, org.jspecify — lower risk but
  no reason to leave unrelocated

After this change, only com.linkedin (Pegasus models) and com.datahub
remain unrelocated, both unique to DataHub with no Spark conflict.

treff7es committed:
fix(ingest/spark): relocate all possible packages, document com.linkedin limitation

Added com.google.thirdparty relocation. Attempted com.linkedin and
com.datahub but Pegasus data framework uses reflection and class name
strings internally for serialization — relocating it causes
DataEncodingException at runtime. Added comments documenting this
constraint.

Final state: only com.linkedin (Pegasus models) and com.datahub remain
unrelocated. Both are unique to DataHub with no Spark runtime conflict.

treff7es committed:
fix(ingest/spark): relocate io.openlineage to prevent classpath conflicts

Relocate io.openlineage to io.acryl.shaded.io.openlineage to prevent
conflicts when another openlineage-spark version is on the Spark
classpath (common in Databricks, EMR, Glue environments).

The shadow plugin relocates both the upstream dependency classes and
DataHub's customized classes to the same shaded package. The override
mechanism (project classes beat dependency classes during JAR merging)
still works because merging happens before relocation. Service loader
files (META-INF/services/) are also rewritten automatically.

Verified with spark-submit: lineage with input, output, and column-level
lineage all working correctly.

treff7es committed:
fix(ingest/spark): update gradle lockfile to resolve openlineage-spark 1.38.0

The lockfile was pinned to 1.33.0 via dependency locking while build.gradle
already declared 1.38.0, causing a version constraint conflict on CI.

askumar27 commented on metadata-integration/java/acryl-spark-lineage/src/main/java/datahub/spark/DatahubSparkListener.java:
```diff
@@ -103,7 +104,7 @@ public void onApplicationStart(SparkListenerApplicationStart applicationStart) {
     this.appContext = getSparkAppContext(applicationStart);
-    initializeContextFactoryIfNotInitialized();
+    initializeContextFactoryIfNotInitialized(applicationStart.appName());
     listener.onApplicationStart(applicationStart);
```
issue (blocking): listener is now lazily initialized, but if `initializeContextFactoryIfNotInitialized()` fails, listener stays null.

treff7es replied:
The listenerNotReady() guard already covers this: every handler calls it right after init and returns early before touching listener, so a failed or disabled init is a clean no-op instead of an NPE back into Spark's bus. That guard was added in 8a45b10 (after this comment), so leaving the lazy init as-is.

askumar27 reviewed (approved):
LGTM! Left some minor comments.

treff7es committed:
fix(ingest/spark): update openlineage-java lockfiles to 1.38.0

treff7es committed:
fix(ingest/spark): restore schemaVersion field accidentally deleted from MetadataChangeProposal avsc

treff7es committed:
fix(patch): restore 2-level array key mapping in GlobalTagsTemplate

GlobalTagsPatchBuilder.addTag() generates paths with a trailing slash
(/tags/<urn>/) which requires 2-level array key handling to correctly
round-trip through transformFields/rebaseFields. The 1-level change
introduced in 15e7f1e580 broke EntityServiceTest patch tests because
the single-level scheme leaves an empty-string key artifact in the
deserialized array entries.

treff7es committed:
Revert "fix(patch): restore 2-level array key mapping in GlobalTagsTemplate"

This reverts commit f0e3b64dc9a7bec5a76b30cdf69d16f8056b0c51.

treff7es committed:
fix(ingest/spark): regenerate lockfiles for openlineage 1.38.0 on current master

The rebase temporarily kept the branch-side lockfiles, which reverted master's
current transitive dependency versions. Regenerated via resolveAndLockAll so the
six openlineage-touching lockfiles carry master's current versions (netty 4.2.15,
httpclient5 5.6.1, etc.) alongside the openlineage 1.38.0 bump.

treff7es committed:
fix(ingest/spark): restore Glue ARN symlink for Iceberg GlueCatalog under OpenLineage 1.38

OpenLineage 1.38 (PR #4053, "fix false hive glue detection") narrowed
AwsUtils.getGlueArn to the Hive-on-Glue metastore case only. That dropped the
arn:aws:glue:<region>:<account> symlink for Iceberg tables read through
GlueCatalog, and GlueCatalogTypeHandler.getIdentifier then wraps the empty
result in Optional.of(null) -- an NPE the plan visitor swallows, so the Iceberg
input dataset is never emitted. This broke SparkGlueConnectionInstanceSmokeTest
(from #17962), which relies on that symlink to resolve the per-connection
platform_instance.

Vendor AwsUtils to backport upstream OL #4384 ("Fix spark iceberg glue catalog
symlink"): detect Iceberg-on-Glue explicitly from spark.sql.catalog.* rather
than reinstating the pre-1.38 blanket fallback (which would reintroduce the
false-positive detection #4053 removed). Remove this vendored file once the
agent upgrades to an OpenLineage release that already contains #4384.

treff7es committed:
feat(ingest/spark): upgrade OpenLineage 1.45.0 -> 1.50.0

Follows the module's documented upgrade process (CLAUDE.md): refresh the vendored
classes onto the upstream 1.50 base, then re-apply DataHub customizations.

- Refresh PathUtils, SaveIntoDataSourceCommandVisitor, WriteToDataSourceV2Visitor to
  the upstream 1.50 base via 3-way merge (base=1.45, ours=1.50, theirs=our custom),
  picking up upstream fixes we would otherwise shadow: BigLake/GCP-lakehouse Hive
  catalog symlinks (1.46), the S3 default-location Path fix (1.47), and the
  DatasetFactory builder-pattern API (1.48). The other 7 tracked files had no upstream
  change 1.45->1.50.
- Migrate remaining DataHub-authored call sites to the builder API
  (getDataset(...) -> sparkDatasetBuilder().dataset(...).schema(...)...build()):
  JdbcSparkUtils, the redshift vendor visitors, FileStreamMicroBatchStreamStrategy,
  and the streaming paths in WriteToDataSourceV2Visitor.
- RddDatasetInfoExtractor: access PartitionedFile.filePath() and
  DataSourceRDDPartition.inputPartition() reflectively in the older-Spark branches so
  they compile against Spark 3.5 (SparkPath / inputPartitions() renames).
- Drop the AwsUtils Glue-ARN backport: OL 1.46 (#4384) ships isIcebergUsingGlue
  natively, verified by SparkGlueConnectionInstanceSmokeTest.
- Regenerate lockfiles and the v1.50.0 customization patches; update CLAUDE.md
  (current version, key changes, known customizations, DataHub-specific files).

Verified: unit tests, the real-Spark smoke suite (incl. the Glue connection-instance
test), and checkShadowJar all pass.

treff7es committed:
test(ingest/spark): add Apache Spark 4.x compatibility smoke test

Adds a sparkSmoke4Test task + src/sparkSmoke4 source set that runs a real local
Apache Spark 4.0.0 job (CSV read -> write) with the DataHub listener attached and
asserts lineage (dataFlowInfo, dataJobInputOutput, the input dataset) is emitted,
confirming the OpenLineage 1.50 plugin supports Spark 4. Wired into the
spark-smoke-test CI workflow so Spark 4 support is guarded going forward.

Spark 4 is Scala 2.13 + Java 17 only, so the test runs against the real shaded
Scala 2.13 agent jar (shadowJar_2_13) where bundled deps are relocated under
io.acryl.shaded (no clash with Spark 4's own antlr/jackson); lives in its own
source set; forces antlr4-runtime 4.13.1 for Spark 4's SqlBaseLexer; aligns the
JUnit platform launcher via the BOM; and opts the test-only configs out of
dependency locking (their deps are not part of the shipped artifact).

Run: ./gradlew :metadata-integration:java:acryl-spark-lineage:sparkSmoke4Test

treff7es committed:
fix(ingest/spark): order jacoco after sparkSmoke4Test + report its results

The spark-smoke CI run failed because jacocoTestReport consumes build/jacoco/*.exec
(now including sparkSmoke4Test.exec) without declaring a dependency on sparkSmoke4Test,
tripping Gradle's implicit-dependency validation. Add mustRunAfter('sparkSmoke4Test')
(matching the existing sparkRealSmokeTest ordering).

Also collect sparkSmoke4Test results in the report-test-results step so Spark 4 test
failures are visible in the CI artifact (previously only test + sparkRealSmokeTest
paths were globbed).

treff7es committed:
fix(ingest/spark): restore spotless lock entries dropped during lockfile regen

An earlier local 'resolveAndLockAll --write-locks' ran in an environment that did
not realize Spotless's dynamically-named configuration (spotless865458226), so the
six openlineage-touching lockfiles lost the google-java-format / guava / errorprone
(etc.) entries that master locks under that config. Restore master's exact entries
for those deps, keeping only the genuine OpenLineage 1.50 dependency changes.

treff7es committed:
docs(ingest/spark): salvage upgrade-research doc into CLAUDE.md + ING-2959, remove doc

The OpenLineage 1.38->1.45 upgrade-research doc was a transient planning artifact (now
superseded by the 1.50 upgrade). Salvage its durable content: the upstream-trimmers vs
DataHub-PathSpec relationship and the column-level-lineage consistency gap are captured as
a new section in CLAUDE.md; the actionable Phase 1/2/3 migration strategy is tracked in
ING-2959. Remove the doc (recoverable from git history).

treff7es committed:
docs(ingest/spark): align upgrade docs + scripts with the real (3-way-merge) process

The documented cp+patch upgrade flow (and upgrade-openlineage.sh) doesn't hold up when
upstream changes a patched file — which OL 1.48's DatasetFactory builder refactor did. Rework
the process to match how the 1.50 upgrade actually worked:

- CLAUDE.md: document the 3-way-merge process (git merge-file: upstream-old -> upstream-new
  delta onto our customized file), the compile-driven fix loop for breaking API changes, and the
  'drop backports upstream now ships' step. Add the lockfile/Spotless gotcha (resolveAndLockAll
  can silently drop the spotless<hash> config — verify and restore).
- CLAUDE.md: replace the stale 'Test thoroughly' step with the real suite — test, checkShadowJar,
  openlineage-converter:test, sparkRealSmokeTest, and the new sparkSmoke4Test — and add a Spark 4
  compatibility note.
- upgrade-openlineage.sh: fix the patch-path bug (looked in the flat dir; patches live under
  v<old-version>/), so its apply step is no longer a silent no-op; add a best-effort warning
  steering to the 3-way merge + test suite.
- Drop the renamed/forked RddPathUtils entry from the fetch/upgrade/generate file lists (upstream
  no longer ships it; it's the DataHub-forked RddDatasetInfoExtractor, ported by hand).

treff7es committed:
fix(ingest/spark): guard lazily-initialized listener against null in event handlers

The delegate OpenLineageSparkListener is created lazily in
initializeContextFactoryIfNotInitialized(), which can return without setting it when the
listener is disabled, there is no active SparkContext/SparkEnv yet, or config parsing
fails. Every event handler then dereferenced 'listener' unconditionally, throwing an NPE
back into Spark's listener bus in exactly those degraded cases.

Add a shared listenerNotReady(SparkListenerEvent) guard so each handler
(onApplicationStart/End, onTaskEnd, onJobStart/End, onOtherEvent) is a clean no-op when
the delegate is unavailable, logging the skipped event by its Spark class name. Add
DatahubSparkListenerNullSafetyTest covering the uninitialized-delegate path.

treff7es committed:
docs(ingest/spark): update README changelog + Spark-version support for OL 1.50

The 'Next' changelog only listed the 1.38/1.45-era changes. Add this PR's headline items:
the OpenLineage 1.50.0 upgrade (with Glue-ARN handling now native upstream), Apache Spark
4.x support, and the listener null-safety fix. Update 'Spark versions supported' to note
Spark 4.x is supported via the Scala 2.13 build.

treff7es committed:
feat(ingest/spark): auto-disable OpenLineage trimmers when DataHub path trimming is configured

OpenLineage 1.39+ enables built-in dataset-name trimmers by default. They run before
DataHub's own path handling (RemovePathPatternUtils / file_partition_regexp) in
OpenLineageRunEventBuilder, so with a DataHub path-trimming mechanism configured they
double-transform names and can break path_spec patterns that reference partition segments.

DatahubSparkListener.configureDatasetTrimmers now, when the user hasn't pinned OpenLineage's
native disabledTrimmers key:
- disables the OL trimmers when a path_spec_list or file_partition_regexp is configured
  (DataHub owns trimming); leaves them on otherwise (upstream default);
- honors an explicit override, spark.datahub.metadata.dataset.openLineageTrimmersEnabled
  (true forces on, false forces off).

Tests:
- DatahubTrimmerDefaultsTest: the resolution matrix (path_spec, file_partition_regexp,
  explicit true/false, native-key-respected) plus a check that the disabled-list class names
  actually match OpenLineage's built-ins.
- SparkTrimmerDisabledSmokeTest: real-Spark regression guard using a deliberately
  non-matching file_partition_regexp, so dt= survives only if the OL trimmers are disabled.
  (The existing SparkPartitionPatternSmokeTest missed the collision because its regex strips
  the same dt= the trimmer does, converging to the same URN.)

Docs: README changelog + CLAUDE.md updated. Deeper integration tracked in ING-2959.

treff7es committed:
docs(ingest/spark): note OL 1.50 trimmer URN change + Spark 4 support in updating-datahub

Add an updating-datahub entry under ## Next:
- Breaking Changes: OpenLineage 1.50's built-in dataset-name trimmers are on by
  default; recipes without path_spec_list/file_partition_regexp now get partition
  segments auto-stripped, changing partitioned file/object-store dataset URNs.
  Documents both opt-outs (configure DataHub path trimming, or set
  openLineageTrimmersEnabled=false).
- Other Notable Changes: Apache Spark 4.x support (Scala 2.13 + Java 17).

treff7es committed:
docs(ingest/spark): fix markdown emphasis style in CLAUDE.md (prettier)

askumar27 commented on metadata-integration/java/acryl-spark-lineage/src/main/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitor.java:
```diff
@@ -142,52 +146,36 @@ public List<OutputDataset> apply(SparkListenerEvent event, SaveIntoDataSourceCom
         .getClass()
         .getCanonicalName()
         .equals(JdbcRelationProvider.class.getCanonicalName())) {
+      if (!command.options().get("dbtable").isDefined()
```
**Undocumented lineage-coverage change (please confirm intentional).** This hunk drops the DataHub-specific catalog-based Delta handling — the `command.options().contains("table")` branch and `extractTableNameFromContext(qe)` (added in the 1.33 bump, #13347) — and replaces it with a `dbtable`/`url` (JDBC) check. After this, a path-less/table-option Delta `saveAsTable` no longer emits an `OutputDataset` from this visitor.

The AwsUtils/Glue backport removal is called out in the description, but this one isn't. Please confirm OL 1.50 now covers it upstream (vs. accidental loss in the 3-way merge), and add a line to `updating-datahub.md` if Delta lineage coverage genuinely moves/drops.

treff7es replied:
Good catch - this was an accidental loss in the 3-way merge, not upstream coverage. The catalog-based handling (the `table` option branch and the CREATE-TABLE-from-SQL fallback) lives on master but was never captured as a patch, so it got dropped when the vendored file moved to 1.50. Restored both branches on the new `sparkDatasetBuilder` API and added them to the v1.50.0 `SaveIntoDataSourceCommandVisitor.patch` so future upgrades keep them. Also made the SQL fallback skip an optional `IF NOT EXISTS` to reach the real name. No `updating-datahub.md` entry since coverage is preserved (restored), not moved/dropped.

askumar27 commented on metadata-integration/java/acryl-spark-lineage/src/main/java/io/openlineage/spark/agent/util/RddDatasetInfoExtractor.java:
```diff
@@ -0,0 +1,371 @@
+                  // filePath returns SparkPath for Spark 3.4
+                  return tryExecuteMethod(f, "filePath")
+                      .map(o -> tryExecuteMethod(o, "toPath"))
+                      .map(o -> (Path) o.get())
```
**Reflective miss throws instead of dropping one dataset.** This Spark 3.4+ branch chains `.map(o -> (Path) o.get()).get()` — both `Optional.get()`s throw `NoSuchElementException` if the reflective `filePath()`/`toPath()` lookup returns empty on an unexpected Spark point-release. The sibling Spark <3.4 branch just below correctly uses `.orElse(null)` and relies on the existing `.filter(Objects::nonNull)`.

That matters because the new `listenerNotReady` guard in `DatahubSparkListener` only guards a *null delegate* — it does not catch exceptions thrown inside `listener.onXxx(...)`. So a reflection shape-mismatch here aborts extraction rather than silently dropping a single dataset. Suggest mirroring the null-safe style: map to `null` and let `filter(Objects::nonNull)` drop it.

treff7es replied:
Fixed - the 3.4+ branch is now `flatMap`/`map`/`orElse(null)` like the sibling `<3.4` branch, so a reflective miss on an unexpected point-release drops that one file via the existing `filter(Objects::nonNull)` instead of throwing NoSuchElementException out of the listener.

askumar27 commented on metadata-integration/java/acryl-spark-lineage/src/main/java/datahub/spark/DatahubSparkListener.java:
```diff
@@ -363,6 +387,86 @@ private static void initializeMetrics(OpenLineageConfig openLineageConfig) {
+  private boolean listenerNotReady(SparkListenerEvent event) {
+    if (listener == null) {
+      log.debug(
+          "OpenLineage listener is not initialized; skipping {}", event.getClass().getSimpleName());
```
**Observability: an inert lineage agent is only visible at DEBUG.** If init fails (the `URISyntaxException` catch logs one narrow ERROR and leaves `listener` null) — or if a non-start event fires when there's no active `SparkEnv` (the no-arg `initializeContextFactoryIfNotInitialized()` → `getContext().ifPresent(...)` silently short-circuits, so `listener` stays null with **zero** log) — the agent emits nothing for the whole application and this line whispers at DEBUG per event.

For a lineage tool, "completely inert" is the worst outcome, and it's invisible at default INFO/WARN log levels. Suggest a one-time WARN (guard with an `AtomicBoolean`) that states the app-level consequence ("DataHub Spark lineage is disabled for this application; no events will be emitted"), rather than only the narrow parse error + per-event DEBUG.

treff7es replied:
Added a one-time WARN (AtomicBoolean-guarded) in listenerNotReady stating the app-level consequence ("lineage disabled for this application; no events will be emitted"). Per-event skips stay at DEBUG.

askumar27 commented on metadata-integration/java/acryl-spark-lineage/src/main/java/datahub/spark/DatahubSparkListener.java:
```diff
@@ -363,6 +387,86 @@ private static void initializeMetrics(OpenLineageConfig openLineageConfig) {
+      disableTrimmers = isDatahubPathTrimmingConfigured(sparkConf);
+    }
+    if (disableTrimmers) {
+      sparkConf.set(OL_DISABLED_TRIMMERS_KEY, OL_DEFAULT_DISABLED_TRIMMERS);
```
**Observability: this load-bearing switch mutates user config silently.** `configureDatasetTrimmers` is the central behavior change of the PR, but no branch logs anything. The auto-detect heuristic (`key.startsWith("spark.datahub.") && key.endsWith(".path_spec_list")`) can mis-fire both ways:

- PathSpec configured under a key shape that doesn't match this suffix → trimmers left **on** → OL trimmer + DataHub PathSpec double-transform the URN → orphaned lineage.
- Any single `path_spec_list` present → trimmers disabled **globally**, silently changing naming for platforms the user didn't configure PathSpec on.

`updating-datahub.md` promises trimmers are "disabled automatically when you configure path trimming" — a detection miss silently breaks that promise, discoverable only by spotting broken URNs after the fact. Please log the chosen branch at INFO (native-key-pinned / explicit-override / auto-disabled-due-to-path-trimming / left-enabled).

treff7es replied:
configureDatasetTrimmers now logs the chosen branch at INFO: honoring an explicit OL disabledTrimmers key / explicit openLineageTrimmersEnabled override / auto-disabled because DataHub path trimming is configured / left enabled. So the config mutation is visible at the default level instead of only discoverable via broken URNs.

askumar27 commented on metadata-integration/java/openlineage-converter/src/main/java/io/datahubproject/openlineage/dataset/DatahubJob.java:
```diff
@@ -207,6 +207,16 @@ private void generateDataJobInputOutputMcp(
+    // UPSERT path would create the aspect with empty arrays. Later PATCH emissions then fail to
+    // override the empty UPSERT, causing output edges to be lost.
+    if (inputEdges.isEmpty() && outputEdges.isEmpty() && parentJobs.isEmpty()) {
+      log.info("Skipping dataJobInputOutput emission for {} - no edges to emit yet", jobUrn);
```
nit/suggestion: the skip is right for early coalesced events, but it's unconditional on emptiness — so in **UPSERT mode** (`!config.isUsePatch()`), where emitting the empty aspect is the only way to clear stale edges, a job that legitimately lost all its edges can no longer be cleared (prior edges stick permanently). Consider gating the skip on `config.isUsePatch()` (only skip when a later PATCH can still populate). The `"...no edges to emit yet"` wording also mis-asserts prematurity in the genuine-clear case.

treff7es replied:
Gated the skip on `config.isUsePatch()`. Checked the original fix (37516e3): the customer ran usePatch=true - the empty emission fell through to the UPSERT branch (the PATCH branch requires a non-empty set) and a later PATCH could not override it. So gating on isUsePatch keeps that fix (PATCH-mode empty still skipped) and lets pure-UPSERT users emit the empty aspect to clear stale edges. Fixed the log wording and updated the test to the true-mode scenario, plus added a UPSERT-clear test.

askumar27 reviewed (approved):
Approving — well-structured vendored upgrade. The shading is correct (`com.linkedin`/`com.datahub`/the `datahub.spark` entry point are intentionally left unrelocated), the `DatahubJob` change correctly stays on the PATCH path (it dedupes a double-`build()` and stops an empty UPSERT from clobbering a later PATCH), and there's real coverage — a live Spark 4.0.0 smoke test against the shaded jar, the trimmer tri-state pinned on both branches, and a null-listener safety test that drives the real handlers.

No blockers. A few non-blocking notes inline; the two I'd most like addressed before merge are **observability on the two silent behavioral switches** (the inert-agent state and the trimmer auto-disable both decide silently at DEBUG/no-log) and **confirming the Delta `saveAsTable` removal is intentional + documented**. Details inline.

_Reviewed against DataHub's Java-backend + ingestion-upgrade guidelines._

treff7es committed:
Merge remote-tracking branch 'origin/master' into ol-1.45-rebase

# Conflicts:
#	docs/how/updating-datahub.md

treff7es committed:
fix(ingest/spark): address OL 1.50 PR review — restore Delta saveAsTable lineage, null-safety & observability

Addresses review feedback on #14911:

- SaveIntoDataSourceCommandVisitor: restore the DataHub catalog-based Delta
  handling (the "table" option branch and the CREATE-TABLE-from-SQL fallback)
  that master carries but was silently dropped in the 3-way OL upgrade merge —
  it was never captured as a patch. Re-added on the new sparkDatasetBuilder API
  and now tracked in the v1.50.0 patch so future upgrades keep it. The SQL
  fallback also now skips an optional IF NOT EXISTS qualifier.
- RddDatasetInfoExtractor: make the Spark 3.4+ reflective filePath/toPath path
  null-safe (map-to-null + filter) instead of chaining Optional.get(), so a
  reflective miss on an unexpected point-release drops one dataset rather than
  throwing NoSuchElementException out of the listener.
- DatahubSparkListener: log once at WARN when the lineage agent stays inert for
  an application (previously only visible at DEBUG); log the chosen dataset-name
  trimmer branch at INFO so the silent config mutation is observable.
- DatahubJob: only skip the empty dataJobInputOutput emission in PATCH mode; in
  UPSERT mode emit it so a job that lost all edges can clear stale ones.

Tests: parseCreateTableName unit tests for the Delta SQL fallback; DatahubJob
UPSERT-clear vs PATCH-skip coverage.

treff7es committed:
Merge remote-tracking branch 'origin/master' into ol-1.45-rebase

# Conflicts:
#	docs/how/updating-datahub.md

treff7es committed:
Merge remote-tracking branch 'origin/master' into ol-1.45-rebase

# Conflicts:
#	docs/how/updating-datahub.md
